Stop cloning arc between execution stages in the demo &

mmandlis opened the pull request:
rename arc's clone method

mmandlis commented on runtime/arc.js:
```diff
@@ -102,11 +102,12 @@ class Arc {
 
-  clone() {
+  // Makes a copy of the arc used for speculative execution.
+  makeSpeculative() {
```
i was debating about the name
- makeSpeculativeClone: seems too long
- cloneSpeculative: reads as if we were cloning an already speculative arc
and in general, we're only cloning some of the fields, so maybe "clone" is not the right word, WDYT?

mmandlis replied:
somewhat related 
https://github.com/PolymerLabs/arcs/blob/596fc4ea726e98a1f1e019489aa5f36dfa566349/runtime/arc.js#L77 (i don't know how to comment directly on that line, if it wasn't change in this PR)

var particle = arc.instantiateParticle(serializedParticle.name);

seems to expect instantiateParticle to return the particle, instead it returns a handle to the particle. we should either update the variable name, or the return value - which one is preferable?

shans replied:
I think makeSpeculativeClone is fine, it's only used in one place.

cloneForSpeculativeExecution would be another option?

shans replied:
It can't return the particle because of the PEC barrier, so renaming might be better. OTOH, this *is* instantiating a particle .. just on the other side of the barrier.

mmandlis replied:
done - cloneForSpeculativeExecution it is

mmandlis commented on runtime/browser/browser-vr-demo/browser-vr-demo.js:
```diff
@@ -121,23 +121,9 @@ suggest(stage++);
 
 let chooseSuggestion = index => {
   document.querySelector('[particle-container]').textContent = '';
-  arc = cloneArc(arc); //arc.clone(); //prepareExtensionArc();
```
Scott, I tried to run the VR demo, but it didn't work (only saw 3 tiny light-grey squares on the top of the page) - do i need to configure/install anything?

sjmiles replied:
The best version of that code is in another branch, I'll try to have it all unified this week.

mmandlis commented on runtime/outer-PEC.js:
```diff
@@ -114,9 +114,11 @@ class OuterPEC extends PEC {
-    this._apiPort.InstantiateParticle(particleSpec, { particleName: particle.name, views })
+    this._apiPort.InstantiateParticle(particleSpec, { particleName: particle.name, views });
+
+    return particleSpec;
```
I'm not sure this is the right thing to return here, but it clearly is expected to return something:
https://github.com/PolymerLabs/arcs/blob/596fc4ea726e98a1f1e019489aa5f36dfa566349/runtime/arc.js#L161

shans replied:
Seems right

shans commented on runtime/arc.js:
```diff
@@ -102,11 +102,12 @@ class Arc {
     var viewMap = new Map();
     this._views.forEach(v => viewMap.set(v, v.clone()));
-    arc.particles = this.particles.map(p => p.clone(viewMap));
+    // TODO: do we need to clone particles?
```
Yes, we absolutely do :)

This is a data structure that tracks particle connections, we don't want any changes in the clone to leak back into the original arc.

mmandlis replied:
changed it to TODO (i'll file an issue to make sure i don't forget to do it).

mmandlis commented:
Thanks for the review. I was in process of addressing the comments when I finally encountered the problem Scott mentioned last week - without cloning the arc in the demo, the GiftList particle is not working properly.
I tracked down the reason to this (in the resolver), and I'll fix it first thing tomorrow morning. Sorry for the delay.

sjmiles commented:
>> without cloning the arc in the demo, the GiftList particle is not working properly

Fwiw, the issue wasn't really in the cloning, it was in controlling the views such that there was only one list view to resolve. The cloning was done primarily to banish the old Particle setup (repeal and replace?) In my DemoWork branch, I'm switching over to mutation instead.

mmandlis commented:
The resolver was implemented in a way that it couldn't match by constraint name the views existing in the Arc, it would only match them by type - so you removing all the views except that ones that you intended to use helped work around this problem. I added a fix to resolver instead.
Another issues was needing to free the slot in SlotComposer (unless the arc is cloned, and SlotComposer is created from scratch) - this will go away when we switch back to suggestion-composer (but I need to rewrite it first).




"Fwiw, the issue wasn't really in the cloning, it was in controlling the views such that there was only one list view to resolve. The cloning was done primarily to banish the old Particle setup (repeal and replace?) In my DemoWork branch, I'm switching over to mutation instead."

mmandlis commented:
Shane, i made additional changes to the resolver in this PR after you reviewed. I'll merge it now, and I'm happy to address any additional comments you might have in follow up PRs. Thanks!

sjmiles commented:
>> another issues was needing to free the slot in SlotComposer

Free what slot?

mmandlis commented:
the "root" slot (when next stage suggestion is chosen, if the arc is not cloned, we continue using the same SlotComposer object, which has the "root" slot not available, so the new recipe's particles cannot be rendered unless the "root" slot is freed)

shans commented:
Rather than using the constraints from the old recipe, would it be better to make use of the new tags in the demo? I think there's an example of them in use in runtime/test/demo-flow-test.js

We really need to implement cloning of particle specs before going any further - what was the problem with the line you deleted and replaced with a TODO?

mmandlis commented:
Shane, 
1) the infrastructure for setting tags is in place, but arc.tagView is never called in the browser-demo (if this is a gap between demo-flow-test and browser-demo, i'll fix it shortly) 

2) "p.clone(viewMap)" refers to a nonexistent clone method. Moreover p is just an object - {particle, views, renderMap, exposeMap}.
This line was only ever working because we were using a new arc with an empty this.particles array for each suggestinator run, so it was never executed.
And even if arc.particles would have been populated, it would be populated with "undefined", because outPec used to not return a value. 
My point is that this changes is not a regression.

sjmiles commented:
re: `root` slot: when we are _mutating arcs_, we never need a new root slot. 

The only thing suggestions can do (at this time) is add particles to open slots.

mmandlis commented:
if you mutate the arc, doesn't the slotComposer still reflect the fact that ListView is occupying the "root" slot, which GiftList wants to render into?
ps: maybe better to discuss this offline?

shans commented:
Yeah, this is a gap - I only implemented the tag mechanism last week so it hasn't had a chance to migrate into the browser stuff yet. Using tags is safer because constraints won't necessarily have names associated with them once we switch to Doug's new stuff.

Point taken re: the clone code being busted :) I think that your point about us only ever using a new arc in the past is also why it's important that we fix this now - we haven't had an opportunity to use the system in a way that would show what will break yet.

mmandlis commented:
Scott, sorry for the misunderstanding - I now found your branch where GiftList stops rendering into the "root" slot, where this hack in the browser-demo will no longer be needed.

Shane, do you think it is worth adding support got TagConnections now, or shall we defer until the new manifest syntax and strategizer implementation is rolled out?

re:cloning, i can send a change with the missing particles cloning implemented shorty. but what I'm wondering is, the not cloning didn't really break anything. 
* arc.particles is not used at all anywhere
* arc.particleViewMaps is populated first thing by recipe.instantiate(arc), so populating it by cloning won't have any affect at all - should it instead be populated by clone and not be populated by the speculative execution?

shans commented:
There's already support - just .tag() the connection specification to further constrain it. See the example in the demo directory.

The existing particleViewMaps shouldn't be repopulated by speculative execution, I don't know how that would happen. I think they're used internally in the PEC to do some stuff (like view routing) but maybe they aren't.

mmandlis commented:
For tags, should it be enough for the connection's tag to match either of the view's tags? should it be an exact match?

If particleViewMaps aren't used, shall i remove them, or keep for future use?
or I could clone particleViewMaps, and not instantiate particle, if it exists in the mapping already (which would exactly handle the speculative execution case).
WDYT?
